torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -232,6 +232,70 @@ Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p
   return self_scaled * scale_v;
 }
 
+Tensor norm_jvp(
+  const Tensor& self_p, const Tensor& self_t,
+  const optional<Scalar> & p_,
+  Tensor norm,
+  IntArrayRef dim,
+  bool keepdim
+) {
+  // NB: currently norm_jvp is also reused for dist's jvp (which haas two differentiable inputs)
+  //     but self_t still cannot be a ZT because that would require both self_t and other_t to be ZT
+  TORCH_INTERNAL_ASSERT(!self_t._is_zerotensor());
+  size_t ndim = self_p.dim();  // composite compliance?
+  double p = p_.value_or(2.0).toDouble();
+
+  if (p == 0.0) {
+    return at::zeros_like(norm);
+  } else if (p == 1.0) {
+    auto result = self_p.sgn();
+    result = areAnyTensorSubclassLike({self_t}) ? result.mul(self_t.conj()) : result.mul_(self_t.conj());
+    result = at::real(result);
+    return result.sum(dim, keepdim);
+  } else if (p == 2.0) {
+    auto result = self_p;
+    result = areAnyTensorSubclassLike({self_t}) ? result.mul(self_t.conj()) : result.mul_(self_t.conj());
+    result = at::real(result);
+    result = result.sum(dim, keepdim);
```

Comment:
It will most likely be more efficient to do a `.dot()` instead of mul+sum.

Reply:
hmm still working on this one. Seems like I need to movedim, unsqueeze, and then do a matmul, squeeze, then unsqueeze the desired dims if keepdim is True.

Reply:
Hmmm not sure?
There might be an extra unsqueeze if you keepdim yes.
If it's more complex you can leave it like that yes.

Reply:
It is pretty complex, so I'm okay with just leaving it. I played around with it on python though and it wasn't faster on any shape, potentially due to python overhead?

```python
import torch

def movedims_to_back_and_flatten(t, dims):
  count = 0
  for dim in reversed(dims):
    count += 1
    t = t.movedim(dim, -count)
  return t.flatten(-count)

def unsqueeze_multiple(t, dims):
  for dim in dims:
    t = t.unsqueeze(dim)
  return t

def mul_then_sum(a, b, dims, keepdim):
  return (a * b).sum(dims, keepdim=keepdim)

def batched_dot(a, b, dims, keepdim):
  a_view = movedims_to_back_and_flatten(a, dims).unsqueeze(-2)
  b_view = movedims_to_back_and_flatten(b, dims).unsqueeze(-1)
  result = (a_view @ b_view).squeeze(-1).squeeze(-1)
  if keepdim:
    result = unsqueeze_multiple(result, dims)
  return result

shape = (2, 4, 5)
for keepdim in (True, False):
  for dims in ((0,), (1,), (0, 1), (0, 2), (1, 2), (0, 1, 2)):
    a = torch.rand(shape)
    b = torch.rand(shape)
    assert torch.allclose(mul_then_sum(a, b, dims, keepdim), batched_dot(a, b, dims, keepdim))
```

```
shape = (512, 256, 256)

a = torch.rand(shape, device="cpu")
b = torch.rand(shape, device="cpu")
dims = (0,)
keepdim = True
```

```
%%timeit

mul_then_sum(a, b, dims, keepdim) # 10 loops, best of 5: 82.6 ms per loop
```

```
%%timeit

batched_dot(a, b, dims, keepdim) # 1 loop, best of 5: 1.41 s per loop

```